### PR TITLE
fix(llmloop): scope async memory compression to each RunPerFile conversation

### LINKED ISSUE #384: [Bug] Async memory compression state is shared across concurrent file reviews
### OpenCodeReview Version

main @ 4ee453f (built from source)

### Operating System

macOS (Apple Silicon)

### Installation Method

Built from source

### LLM Provider

Other OpenAI-compatible endpoint

### Bug Description

When multiple files are reviewed concurrently, `llmloop.Runner` shares one asynchronous memory-compression job across all `RunPerFile` conversations.

The runner is session-scoped, but `pendingJob` is conversation-scoped state:

```go
type Runner struct {
    // ...
    compressionMu sync.Mutex
    pendingJob    *compressionJob
}
```

`addNextMessage` calls `tryApplyPendingCompression(messages)` and `cancelPendingCompression()` without associating the job with the current file or conversation. With `--concurrency > 1`, a compression job started for file A can therefore be observed, applied, replaced, or canceled by file B.

There is also a same-conversation cancellation path: async compression can start when the pre-append token count exceeds the 60% soft limit, then the newly appended assistant/tool-result messages can push the conversation beyond the 80% warning limit in the same `addNextMessage` call, immediately canceling the request that was just started.

This produces `context canceled` errors for memory-compression requests. Some OpenAI-compatible gateways surface that client-side cancellation as HTTP 500, even though ordinary review requests continue to succeed. More importantly, applying a summary created from another file's message history can affect review correctness.

### Steps to Reproduce

1. Configure a review template with `MEMORY_COMPRESSION_TASK` and a sufficiently small `MAX_TOKENS` so the thresholds are reached predictably.
2. Start a review containing at least two files with `--concurrency 2` or higher.
3. Let file A cross the 60% soft limit and keep its compression request in flight.
4. Continue file B until it calls `tryApplyPendingCompression` or crosses the 80% warning limit.
5. Observe that the single `Runner.pendingJob` can be applied or canceled by file B, and the compression client receives `context canceled`.

A deterministic regression test can use a blocking fake LLM client:
- start compression for `a.go`;
- enter `addNextMessage` for `b.go`;
- assert that B neither cancels nor applies A's compression result.

A separate same-file case can start compression just above the soft limit and append enough messages to cross the warning limit; the current ordering starts and then immediately cancels the job.

### Expected Behavior

Each `RunPerFile` conversation should own its compression state.

- A compression summary must only be applied to the message history from which it was created.
- One file must not cancel, replace, or consume another file's compression job.
- The threshold handling should not launch an async request that is immediately canceled in the same message update.
- Aggregate token counters and warnings may remain shared at the runner/session level.

### Logs / Error Output

```shell
memory_compression_task: context canceled
POST [redacted OpenAI-compatible endpoint]: context canceled

In one anonymized run with concurrency 4, ordinary LLM requests continued to succeed while 4 of 12 memory-compression requests were canceled. One cancellation happened within a few milliseconds, consistent with a soft-limit start followed by warning-limit cancellation in the same update path.
```

### Additional Context

Relevant code on current main:

- [Runner stores one shared pending job](https://github.com/alibaba/open-code-review/blob/4ee453f/internal/llmloop/loop.go#L41-L56)
- [addNextMessage applies/cancels the shared job](https://github.com/alibaba/open-code-review/blob/4ee453f/internal/llmloop/loop.go#L433-L469)
- [async compression lifecycle](https://github.com/alibaba/open-code-review/blob/4ee453f/internal/llmloop/compression.go#L255-L332)

Suggested fix: move pending compression ownership into per-`RunPerFile` state, or key the job by a stable conversation/file identifier and validate ownership before applying or canceling it. Also decide whether to start async compression only after appending the new messages, so a job is not started and canceled in the same update.

A conservative alternative is synchronous compression at the threshold, trading some latency for a simpler lifecycle.

This is not a request to retry `context canceled`. Once the parent context is canceled, retrying the same request context cannot succeed; the issue is the ownership and lifecycle of the compression request.

I searched the existing issues and did not find a report for this shared-state race.

### PR DESCRIPTION
**Pull request by chethanuk**
# fix(llmloop): scope async memory compression to each RunPerFile conversation

## Problem

One `Runner` serves a whole review session, and both `internal/agent` and `internal/scan` fan out one goroutine per file over it (default concurrency 8). Async memory compression state, however, lived on the Runner as a single shared slot (`compressionMu` + `pendingJob`). As analyzed in #384 (and confirmed by the maintainer), that shared slot breaks per-conversation semantics once concurrency > 1.

## Root cause — four concrete defects

1. **Cross-file apply** — `tryApplyPendingCompression` had no owner check: file B could consume file A's completed job, splicing `[frozen + A's summary + A's active rounds]` into **B's** message history. A review-correctness bug, not just wasted work.
2. **Cross-file cancel / replace** — the warning-threshold paths canceled whichever file's job happened to be pending, so another file's in-flight compression request died with `context canceled` (rendered as an HTTP 500 by some gateways); `triggerAsyncCompression` also overwrote the slot unconditionally, burning the superseded request.
3. **Same-call start-then-cancel** — the soft-threshold trigger ran **before** the new assistant/tool messages were appended. If the append crossed the warning threshold, the same `addNextMessage` call canceled the job it had started microseconds earlier (the cancellation-within-milliseconds seen in the issue).
4. **Data race** — the `pendingJob == nil` fast-path read in `addNextMessage` was unlocked.

## Fix

- New `compressionState { mu sync.Mutex; pendingJob *compressionJob }`, created per `RunPerFile` call and threaded through `triggerAsyncCompression` / `tryApplyPendingCompression` / `cancelPendingCompression`. Isolation holds by construction: a conversation physically cannot see another conversation's job. The state is GC'd with the call — no cleanup maps, no keying by file path (which would collide if the same path is reviewed twice).
- The nil-pending gate moved from the caller's unlocked read into `triggerAsyncCompression` as an atomic check-and-set under `st.mu` (also fixes 4).
- The async trigger moved to the **end** of `addNextMessage`, gated on the post-append count sitting strictly between the soft (60%) and warning (80%) thresholds — start-then-cancel within one update is now structurally impossible, and no job starts when the call is about to return false (also fixes 3).
- `RunPerFile` defers `cancelPendingCompression(st)`, so an in-flight job is aborted when its conversation ends instead of running up to 5 minutes orphaned.
- Aggregate token counters, warnings, and tool-call counts stay Runner-level, exactly as before. No exported API changes; `internal/agent` and `internal/scan` are untouched.
- Failure reporting follows ownership: the `[ocr] Memory compression failed` line is printed by the sync call sites and by the async job only while it still owns the slot. A **deliberately** cancelled job — superseded by sync compression at the warning threshold, or aborted by `RunPerFile`'s deferred cancel at conversation end — dies silently (its session task record is still written, so nothing is hidden from debugging). Without this, the fix's own intentional cancels would keep printing the issue's exact `context canceled` symptom whenever a job was still in flight at the warning threshold or when the model called `task_done` a round after a trigger.

**Intentional behavior deltas** (all strictly safer): the async snapshot now includes the just-appended round (strictly more history summarized); in-flight jobs are canceled at conversation end; no async job starts on a failing update.

## Tests

New regression tests (channel-synchronized, no sleeps):

| Test | What it proves |
|---|---|
| `TestCompression_CrossFileIsolation` | With file A's compression request held in flight by a gated fake client, file B's cancel + apply touch nothing: A's job stays pending, zero requests canceled (defect 2); also pins that a second trigger while a job is pending is a no-op, not a replacement |
| `└ SummaryAppliesOnlyToOwningConversation` (subtest) | After A's job completes, B cannot consume it and B's messages stay byte-identical; A applies its own summary and a message appended after the snapshot survives (defect 1) |
| `TestAddNextMessage_NoStartThenCancelSameCall` | Pre-append count in (soft, warn), post-append over warn — the #384 same-file case: exactly one sync compression, zero mid-flight cancellations, no dangling async job (defect 3) |
| `TestRunPerFile_ConcurrentFilesCompression_Race` | 4 concurrent `RunPerFile` calls on one Runner crossing both thresholds repeatedly; asserts `-race` cleanliness end to end (defect 4), and asserts compression requests were actually issued so the test can never silently go vacuous |

The RED baseline was demonstrated on unmodified `main`: a gated-client test showed file B's cancel destroying file A's pending job (`file A's job should still be pending after file B's cancel`).

All seven existing compression unit tests were updated mechanically to the per-conversation API; the remaining compression edge cases (empty template no-op, LLM error → messages unchanged, empty summary → unchanged, snapshot-suffix preservation, fence stripping) are unchanged and still green.

```
go test ./internal/llmloop/ -race -count=1   # green
make check && make test && make coverage      # green, coverage 81.1% (≥ 80% gate)
go build ./cmd/opencodereview                 # green
```

## E2E

Mirroring the issue's setup: 2-file review at `--concurrency 2` against a local OpenAI-compatible stub, with responses sized so each conversation crosses the soft threshold (async compression) and the warning threshold (sync compression). Result: both files `review_item_done`, `memory_compression_task` records present for each file, and **zero** `context canceled` / `Memory compression failed` in stdout or session records.

Fixes #384

### CODE REVIEW

File: internal/llmloop/compression.go [hunk 1]
```diff
@@ -252,31 +263,36 @@ func (r *Runner) runCompression(ctx context.Context, msgs []llm.Message, filePat
...
 	msgSnapshot := copyMessages(messages)
-
 	asyncCtx, cancel := context.WithTimeout(context.WithoutCancel(ctx), 5*time.Minute)
-
 	job := &compressionJob{done: make(chan struct{}), cancel: cancel, snapshotLen: len(messages)}
-	r.compressionMu.Lock()
-	r.pendingJob = job
-	r.compressionMu.Unlock()
+	st.pendingJob = job
+	st.mu.Unlock()
```
[the remark this answers is not in the document]
Comment (chethanuk) [reply internal/llmloop/compression.go:279]: Intentional, for two reasons:

1. The nil check itself must be under `st.mu` — the async goroutine writes `st.pendingJob = nil` on its failure path, so an unlocked read here is a data race (it's exactly defect 4 from #384 that this PR removes; the old pre-lock ordering existed *because* of that unlocked read). Moving the copy back out while keeping the check locked would need an unlock→copy→relock double-check, which is more code for no gain.

2. There is no contention to minimize: this branch only runs when `st.pendingJob == nil`, i.e. no async goroutine exists for this conversation yet, and `triggerAsyncCompression` is only ever called by the conversation's own goroutine (single-threaded per `compressionState`). Nothing can block on `st.mu` while the copy runs.

Happy to restructure if a maintainer wants change..

**Review by lizhengfeng101 — COMMENTED** [review]
## Review

Excellent concurrency bug fix. The root cause analysis is thorough (four distinct defects cleanly decomposed), the fix eliminates the problem structurally (per-conversation isolation rather than more locks), and the tests are well-designed with channel synchronization instead of sleeps. A few minor nits below, none blocking.

### 1. `copyMessages` runs under `st.mu` in `triggerAsyncCompression`

```go
st.mu.Lock()
if st.pendingJob != nil { st.mu.Unlock(); return }
msgSnapshot := copyMessages(messages)  // still holding st.mu
...
st.mu.Unlock()
```

Since `st` is per-conversation, the only contender is the background goroutine's completion handler, so there's no real performance concern. But `copyMessages` is unrelated to the mutex-protected state (`pendingJob`) — it could be moved outside the critical section (check-and-set a sentinel under the lock, unlock, copy, re-lock to install the real job). That said, the current approach is safe and simpler. Just a note, not a blocker.

### 2. Comment describes lock ordering that doesn't match the code

In the async goroutine:

```go
// cancelPendingCompression clears pendingJob before cancelling
```

But the actual code in `cancelPendingCompression` calls `cancel()` first, then sets `st.pendingJob = nil`. Both happen under `st.mu`, so the goroutine can never observe the intermediate state — behavior is correct. The comment just describes the wrong order. Suggest rewording to something like "cancels and clears pendingJob under the lock".

### 3. Consider a brief note on the pre-append compression failure path

In `addNextMessage`, when pre-append sync compression fails, execution continues (appends new messages, may retry via the post-append check). This is the right degradation behavior — and an improvement over the old code which silently ignored the error (`_, _ = r.runCompression`). A one-line comment like "// compression failed; continue with over-limit messages, post-append will retry" would make the intent more discoverable, but this is optional.

### Summary

Low-risk, structurally sound fix. The per-conversation `compressionState` eliminates cross-file interference by construction. Test coverage is thorough — `gatedLLMClient` is a clean pattern, the cross-file isolation test pins all four defects, and the race test guards against regression. LGTM.

**Comment by chethanuk** [comment]
Thanks for the thorough review @lizhengfeng101! Items 2 and 3 fixed in aa08bf4: the async-goroutine comment now reads "cancels and clears pendingJob under the lock" (matching the actual order), and `addNextMessage` notes that a pre-append compression failure continues with over-limit messages and is retried by the post-append check.

For item 1 (`copyMessages` under `st.mu`), kept as-is per the inline-thread discussion: the check-first order avoids paying a full-conversation copy on every soft-zone round while a job is already pending (the bot's restructure copies before the pending check), the nil-pending read must stay under `st.mu` anyway (the unlocked read was defect 4 of #384), and there's no contention to shorten — only the conversation's own goroutine calls the trigger. Happy to restructure if you feel strongly.

**Comment by chethanuk** [comment]
cc: @lizhengfeng101

**Review by lizhengfeng101 — APPROVED** [review]
LGTM
